pkg/docker/config/config.go

```diff
@@ -0,0 +1,271 @@
+package config
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/image/types"
+	helperclient "github.com/docker/docker-credential-helpers/client"
+	"github.com/docker/docker/pkg/homedir"
+	"github.com/pkg/errors"
+)
+
+type dockerAuthConfig struct {
+	Auth string `json:"auth,omitempty"`
+}
+
+type dockerConfigFile struct {
+	AuthConfigs map[string]dockerAuthConfig `json:"auths"`
+	CredHelpers map[string]string           `json:"credHelpers,omitempty"`
+}
+
+const (
+	authCfg           = ".config/containers"
+	authCfgFileName   = "auth.json"
+	dockerCfg         = ".docker"
+	dockerCfgFileName = "config.json"
+)
+
+var (
+	// ErrNotLoggedIn is returned for users not logged into a registry
+	// that they are trying to logout of
+	ErrNotLoggedIn = errors.Errorf("not logged in")
+)
+
+// SetAuthentication stores the base64 encoding of username and password
+// in the auth file if the credentials are authenticated
+func SetAuthentication(username, password, registry string) error {
+	path, err := getPathToAuth()
+	if err != nil {
+		return errors.Wrapf(err, "error getting path to auth file")
+	}
+	encodedCreds := encodeUserAndPass(username, password)
+	return addAuthToJSON(registry, encodedCreds, path)
+}
+
+// GetAuthentication returns the value of the registry url stored in
+// the auth.json file
+func GetAuthentication(ctx *types.SystemContext, registry string) (string, string, error) {
+	if ctx != nil && ctx.DockerAuthConfig != nil {
+		return ctx.DockerAuthConfig.Username, ctx.DockerAuthConfig.Password, nil
+	}
```

Comment:
Very unsure about this… on one hand it’s clearly a way to override exactly this data, OTOH it is a bit surprising to have a dedicated ”read config for $registry” facility which is entirely overridden by the system context; usually the system context serves to override a smaller detail of a bigger operation.

When the other functions here don’t use a `types.SystemContext`, it would be more symmetrical not to use it here either; admittedly that’s a fairly weak argument, they might all get a `types.SystemContext` in the future to support overriding or influencing `getPathToAuth`.

Reply:
Before I moved things around, the `getAuth` function in `docker_client.go` had this. So just to not mess with the already established functionality, I moved it here. 

Would it be better is I create a `getAuth` function in `docker_client.go` that checks for the auth using the systemContext, and if none is found it calls `config.GetAuthentication`.
This way the functions will be more symmetrical, unless you see another way around this.

Reply:
`getAuth` was package-private, with a single caller, just to keep the code in smaller units. The public interface now has a somewhat different goal.

Overall, yes, I slightly prefer keeping the `types.SystemContext` overrides in the c/i/docker subpackage. (Either in a separate helper, as you suggest, or directly inlined in the constructor-wrapping function, whichever ends up being cleaner.) But, as I said, I’m very unsure about where the `SystemContext` override should be.

@runcom WDYT?

Reply:
Created `getAuth` in `docker_client.go` that checks the systemContext first then calls `config.GetAuthentication`